src/jdk.jfr/share/classes/jdk/jfr/internal/EventInstrumentation.java

```diff
@@ -90,65 +89,86 @@ public FieldInfo(String fieldName, String fieldDescriptor, String internalClassN
+    private static final Type TYPE_OBJECT = Type.getType(Object.class);
     private static final Method METHOD_COMMIT = new Method("commit", Type.VOID_TYPE, new Type[0]);
     private static final Method METHOD_BEGIN = new Method("begin", Type.VOID_TYPE, new Type[0]);
     private static final Method METHOD_END = new Method("end", Type.VOID_TYPE, new Type[0]);
     private static final Method METHOD_IS_ENABLED = new Method("isEnabled", Type.BOOLEAN_TYPE, new Type[0]);
     private static final Method METHOD_TIME_STAMP = new Method("timestamp", Type.LONG_TYPE, new Type[0]);
+    private static final Method METHOD_GET_EVENT_WRITER_KEY = new Method("getEventWriter", TYPE_EVENT_WRITER, new Type[] { Type.LONG_TYPE });
     private static final Method METHOD_EVENT_SHOULD_COMMIT = new Method("shouldCommit", Type.BOOLEAN_TYPE, new Type[0]);
-    private static final Method METHOD_EVENT_HANDLER_SHOULD_COMMIT = new Method("shouldCommit", Type.BOOLEAN_TYPE, new Type[] { Type.LONG_TYPE });
+    private static final Method METHOD_EVENT_CONFIGURATION_SHOULD_COMMIT = new Method("shouldCommit", Type.BOOLEAN_TYPE, new Type[] { Type.LONG_TYPE });
+    private static final Method METHOD_EVENT_CONFIGURATION_GET_SETTING = new Method("getSetting", TYPE_SETTING_CONTROL, new Type[] { Type.INT_TYPE });
     private static final Method METHOD_DURATION = new Method("duration", Type.LONG_TYPE, new Type[] { Type.LONG_TYPE });
+    private static final Method METHOD_RESET = new Method("reset", "()V");
 
     private final ClassNode classNode;
     private final List<SettingInfo> settingInfos;
     private final List<FieldInfo> fieldInfos;;
-    private final Method writeMethod;
-    private final String eventHandlerXInternalName;
     private final String eventName;
-    private final boolean untypedEventHandler;
-    private boolean guardHandlerReference;
-    private Class<?> superClass;
-
-    EventInstrumentation(Class<?> superClass, byte[] bytes, long id) {
+    private final Class<?> superClass;
+    private final boolean untypedEventConfiguration;
+    private final Method staticCommitMethod;
+    private final long eventTypeId;
+    private boolean guardEventConfiguration;
+
+    EventInstrumentation(Class<?> superClass, byte[] bytes, long id, boolean bootClass) {
+        this.eventTypeId = id;
         this.superClass = superClass;
         this.classNode = createClassNode(bytes);
         this.settingInfos = buildSettingInfos(superClass, classNode);
         this.fieldInfos = buildFieldInfos(superClass, classNode);
-        this.untypedEventHandler = hasUntypedHandler();
-        this.writeMethod = makeWriteMethod(fieldInfos);
-        this.eventHandlerXInternalName = ASMToolkit.getInternalName(EventHandlerCreator.makeEventHandlerName(id));
-        String n =  annotationValue(classNode, ANNOTATION_TYPE_NAME.getDescriptor(), String.class);
+        String n = annotationValue(classNode, ANNOTATION_TYPE_NAME.getDescriptor(), String.class);
```

Comment:
As well - the descriptor could be cached to avoid calling substring operations repeatedly.

Reply:
Changed from Type to String in some places to reduce allocation. More could be done with the Method descriptor, but I didn't want to go overboard. If needed, I rather handle it separately.